streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java

```diff
@@ -0,0 +1,237 @@
+                streamsMetrics
+            );
+
+            final Task task = new StreamTask(
+                taskId,
+                partitions,
+                topology,
+                consumer,
+                config,
+                streamsMetrics,
+                stateDirectory,
+                cache,
+                time,
+                stateManager,
+                recordCollector
+            );
+
+            log.trace("Created task {} with assigned partitions {}", taskId, partitions);
+            createdTasks.add(task);
+            createTaskSensor.record();
+        }
+        return createdTasks;
+    }
+
+    public void releaseProducer() {
+        if (threadProducer != null) {
+            try {
+                threadProducer.close();
+            } catch (final RuntimeException e) {
+                log.error("Failed to close producer due to the following error:", e);
+            }
+        }
+        if (!taskProducers.isEmpty()) {
+            throw new IllegalStateException("Expected task producers to have been cleared before closing");
+        }
+    }
+
+    void releaseProducer(final TaskId id) {
+        final Producer<byte[], byte[]> producer = taskProducers.remove(id);
+        if (producer != null) {
+            try {
+                producer.close();
+            } catch (final RuntimeException e) {
+                log.error("Failed to close producer due to the following error:", e);
+            }
+        }
+    }
+
+    Map<MetricName, Metric> producerMetrics() {
+        final Map<MetricName, Metric> result = new LinkedHashMap<>();
+        if (threadProducer != null) {
+            final Map<MetricName, ? extends Metric> producerMetrics = threadProducer.metrics();
+            if (producerMetrics != null) {
+                result.putAll(producerMetrics);
+            }
+        } else {
+            // When EOS is turned on, each task will have its own producer client
+            // and the producer object passed in here will be null. We would then iterate through
+            // all the active tasks and add their metrics to the output metrics map.
+            for (final Map.Entry<TaskId, Producer<byte[], byte[]>> entry : taskProducers.entrySet()) {
+                final Map<MetricName, ? extends Metric> taskProducerMetrics = entry.getValue().metrics();
+                result.putAll(taskProducerMetrics);
+            }
+        }
+        return result;
+    }
+
+    Set<String> producerClientIds() {
+        if (threadProducer != null) {
+            return Collections.singleton(getThreadProducerClientId(threadId));
+        } else {
+            return taskProducers.keySet()
+                                .stream()
+                                .map(taskId -> getTaskProducerClientId(threadId, taskId))
+                                .collect(Collectors.toSet());
+        }
+    }
+
+    public InternalTopologyBuilder builder() {
```

Comment:
Does this need to be public?

Reply:
Hah, doesn't need to be there at all, actually.